core/src/test/java/com/google/instrumentation/common/ProviderTest.java

```diff
@@ -15,47 +15,111 @@
 
 import static com.google.common.truth.Truth.assertThat;
 
+import java.util.ServiceConfigurationError;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
-/**
- * Tests for {@link Provider}
- */
+/** Tests for {@link Provider} */
 @RunWith(JUnit4.class)
 public class ProviderTest {
+  static class GoodClass {
+    public GoodClass() {}
+  }
+
+  static class PrivateConstructorClass {
+    private PrivateConstructorClass() {}
+  }
+
+  static class NoDefaultConstructorClass {
+    public NoDefaultConstructorClass(int arg) {}
+  }
+
+  private static class PrivateClass {}
+
+  static interface MyInterface {}
+
+  static class MyInterfaceImpl implements MyInterface {
+    public MyInterfaceImpl() {}
+  }
+
   @Test
   public void testGoodClass() throws Exception {
-    GetGen getGen0 =
-        Provider.newInstance("com.google.instrumentation.common.ProviderTest$GetGen", null);
-    assertThat(getGen0).isNotNull();
-    assertThat(getGen0.getGen()).isEqualTo(0);
-    for (int i = 1; i < 10; i++) {
-      GetGen getGenI =
-          Provider.newInstance("com.google.instrumentation.common.ProviderTest$GetGen", null);
-      assertThat(getGenI).isNotNull();
-      assertThat(getGenI.getGen()).isEqualTo(i);
-    }
-    assertThat(getGen0.getGen()).isEqualTo(0);
+    assertThat(
+            Provider.<GoodClass>newInstance(
```

Comment:
Yes.
Reasons are:
* We need to call Class.forName with a literal otherwise shading tools will not be able to replace the path.
* Android class-loader http://stackoverflow.com/questions/13407006